File: day-3/python/wire.py
"""Structures relating to the wire and points on it."""
import typing
from dataclasses import dataclass


@dataclass
class Point:
    """Class representing a point in 2D space."""

    x: int
    y: int

    def right(self: "Point") -> "Point":
        """Return a new point 1 unit right of the point."""
        return Point(self.x + 1, self.y)

    def left(self: "Point") -> "Point":
        """Return a new point 1 unit left of the point."""
        return Point(self.x - 1, self.y)

    def up(self: "Point") -> "Point":
        """Return a new point 1 unit up from the point."""
        return Point(self.x, self.y + 1)

    def down(self: "Point") -> "Point":
        """Return a new point 1 unit down from the point."""
        return Point(self.x, self.y - 1)

    def __hash__(self: "Point") -> int:
        """Convert the Point to a form which Python can hash."""
        return hash((self.x, self.y))

    def manhattan_distance(self: "Point", other: "Point") -> int:
        """Calculate the manhattan distance between two points."""
        return abs(self.x - other.x) + abs(self.y - other.y)


class Wire:
    """Class reprensenting an electrical wire."""

    def __init__(self: "Wire", points: typing.List[Point]) -> None:
        """Instantiate a new wire given a list of points."""
        self.points = points

    def dist_to(self: "Wire", point: Point) -> int:
        """Calculate how far along the wire this point is."""
        return self.points.index(point)

File: day-3/rust/src/wire.rs
use std::collections::HashSet;
use std::convert::TryInto;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Point {
    pub x: i64,
    pub y: i64
}

#[derive(Clone, Debug)]
pub struct Wire {
    points: Vec<Point>
}

impl Wire {
    pub fn new(points: Vec<Point>) -> Wire {
        Wire {
            points
        }
    }

    pub fn to_set(self) -> HashSet<Point> {
        self.points.into_iter().collect()
    }

    pub fn find_steps_to(self, point: &Point) -> i64 {
        self.points.iter().position(|p| p == point).unwrap().try_into().unwrap()
    }
}

impl Point {
    pub fn new(x: i64, y: i64) -> Point {
        Point {
            x, y
        }
    }

    pub fn up(&mut self, distance: i64) {
        self.y += distance;
    }

    pub fn down(&mut self, distance: i64) {
        self.y -= distance;
    }

    pub fn right(&mut self, distance: i64) {
        self.x += distance;
    }

    pub fn left(&mut self, distance: i64) {
        self.x -= distance;
    }

    pub fn manhattan_distance(&self, other_point: &Point) -> i64 {
        (self.x - other_point.x).abs() + (self.y - other_point.y).abs()
    }
}
